### ISSUE TITLE
Converting models to Caffe.

### ISSUE THREAD

**Issue by omkarparkhi** [facebook/fb.resnet.torch#21]
Hi,

I am trying to convert models to Caffe using `https://github.com/facebook/fb-caffe-exts`.

While folding b-norm layers into their predecessing conv layers, this repo looks for variable bn_layer.running_std in a b-norm layer. While models trained using the code here, have bn_layer.running_var

Can you help with these conversions? Should I just converted var into std?

I see that `running_var = running_std:pow(-2):add(-self.eps)` should I just reverse this to get the std and go ahead with the conversions?

**Comment by colesbury** [comment]
Yes, you can reverse it:

`running_std = running_var:add(self.eps):pow(0.5)`

Or even better, add support for `running_var` to `fb-caffe-exts` :smiley:

**Comment by omkarparkhi** [comment]
Thanks. Ok. Let me try it first. If it works, we can think about a pull request on fb-caffe-exts :)

**Comment by omkarparkhi** [comment]
Sam, another quick question, how did you plot the training error curves? I couldn't find that being saved anywhere.

**Comment by colesbury** [comment]
I log a bunch of stats in JSON format and then plot using ipython and bokeh:

Around https://github.com/facebook/fb.resnet.torch/blob/master/main.lua#L61:
```lua
my_json_log(epoch, opt, trainer.optimState, {
      train_err = trainTop1,
      train_err5 = trainTop5,
      train_accuracy = 100 - trainTop1,
      train_loss = trainLoss,
      test_err = testTop1,
      test_err5 = testTop5,
      test_accuracy = 100 - testTop1,
      best_err = bestTop1,
      best_err5 = bestTop5,
      best_accuracy = 100 - bestTop1,
      finished = false,
   })
```

I'm not sure how the Torch community typically plots these curves. If there's a good way, I'd be glad to support it directly in this project.

**Comment by omkarparkhi** [comment]
May be @soumith, knows about torch plotting, but atleast I can save things using your input. Thanks!!

**Comment by szagoruyko** [comment]
The best way is to parse the log file in ipython notebook converting to pandas frame, then it's trivial to visualize it with bokeh. I can post an example if there is interest.

**Comment by omkarparkhi** [comment]
Sure. Please do.

**Comment by omkarparkhi** [comment]
Sam, have you tried converting a model to caffe? I fixed the std/var problem but there are still errors in the code I am not able to get around.

**Comment by colesbury** [comment]
No, I haven't tried converting the models to caffe

**Comment by miquelmarti** [comment]
@omkarparkhi Did you finally have any luck with that?

I want to convert ResNet-18 and Resnet-34.
There is a [tool](https://github.com/jay-mahadeokar/pynetbuilder) to create the prototxt files for ResNets  but I am not sure they will match exactly the models here and I don't want to lose too much time tinkering around with the weights and I really want to avoid training the models from scratch.

**Comment by KevinYuk** [comment]
@miquelmarti 
Hi miquelmarti, 

How about the tool from jay-mahadeokar/pynetbuilder? Have you successfully got the good accuracy by that tool?

Thanks.
